packages/core/src/scuttle.js

```diff
@@ -0,0 +1,185 @@
+      )
+    }
+    opts.scuttlerFunc = globalRef[opts.scuttlerName]
+  }
+  return opts
+
+  /**
+   * @param {string|RegExp} except - Exception to convert to RegExp.
+   * @returns {string|RegExp} - Converted exception.
+   */
+  function toRE(except) {
+    // turn scuttleGlobalThis.exceptions regexes strings to actual regexes
+    if (!except.startsWith('/')) {
+      return except
+    }
+    const parts = except.split('/')
+    const pattern = parts.slice(1, -1).join('/')
+    const flags = parts[parts.length - 1]
+    return new RegExp(pattern, flags)
+  }
+}
+
+/**
+ * Runs scuttling on the globalRef. Use applyDefaultScuttling for full scope of options.
+ *
+ * @param {GlobalRef} globalRef - Reference to the global object.
+ * @param {Array<string|RegExp>} extraPropsToAvoid - List of additional properties to exclude from scuttling beyond the default ones.
+ */
+function performScuttleGlobalThis(globalRef, extraPropsToAvoid = new Array()) {
+  const props = new Array()
+  getPrototypeChain(globalRef).forEach((proto) =>
+    props.push(...getOwnPropertyNames(proto))
+  )
+
+  // support LM,SES exported APIs and polyfills
+  const avoidForLavaMoatCompatibility = ['Compartment', 'Error', 'globalThis']
+  const propsToAvoid = new Set([
+    ...avoidForLavaMoatCompatibility,
+    ...extraPropsToAvoid,
+  ])
+
+  const obj = create(null)
+  for (const prop of props) {
+    // eslint-disable-next-line no-inner-declarations
+    function set() {
+      console.warn(
+        `LavaMoat - property "${prop}" of globalThis cannot be set under scuttling mode. ` +
+          'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/360.'
+      )
+    }
+    // eslint-disable-next-line no-inner-declarations
+    function get() {
+      throw new Error(
+        `LavaMoat - property "${prop}" of globalThis is inaccessible under scuttling mode. ` +
+          'To learn more visit https://github.com/LavaMoat/LavaMoat/pull/360.'
+      )
+    }
+    if (shouldAvoidProp(propsToAvoid, prop)) {
+      continue
+    }
+    let desc = getOwnPropertyDescriptor(globalRef, prop)
+    if (desc?.configurable === true) {
+      desc = { configurable: false, set, get }
+    } else if (desc?.writable === true) {
+      const p = new Proxy(obj, { getPrototypeOf: get, get, set })
+      desc = { configurable: false, writable: false, value: p }
+    } else {
+      continue
+    }
+    defineProperty(globalRef, prop, desc)
+  }
+}
+
+/**
+ * @param {Set<string|RegExp>} propsToAvoid - List of properties to exclude from scuttling.
+ * @param {string} prop - Property to check.
+ * @returns {boolean} - Whether the property should be avoided or not.
+ */
+function shouldAvoidProp(propsToAvoid, prop) {
+  for (const avoid of propsToAvoid) {
+    if (avoid instanceof RegExp && avoid.test(prop)) {
+      return true
+    }
+    if (propsToAvoid.has(prop)) {
+      return true
+    }
+  }
+  return false
```

Comment:
The `Set` is useful to avoid duplicate `string`s, but has no upside if all of the members are `RegExp` objects.  Two separate `RegExp` objects are obviously not guaranteed to not test `true` for the same string.  But the function is kind of difficult to understand as written.  This one is shorter and probably slower: 

```suggestion
  return [...propsToAvoid].some(avoid => new RegExp(avoid).test(prop))
```

don't really care what you use instead as long as it's easier to understand.

Reply:
TBF I find the original code easier to understand than the suggestion. And I wanted to focus on refactoring this. Would you mind if I left it as-is?

Reply:
But would that work for when `avoid` is a string rather than an RE? Maybe:

`propsToAvoid.some(avoid => avoid === prop || new RegExp(avoid).test(prop))`

Reply:
I now see I partially misunderstood the original suggestion. 
@weizman Would you like to take over this PR as you seem more ready to jump right in than I am. Let's bring that to our next pairing session with any changes you see fit.